src/fsharp/CompileOps.fs

```diff
@@ -4291,22 +4292,29 @@ type TcImports(tcConfigP:TcConfigProvider, initialResolutions:TcAssemblyResoluti
                       with e -> error(Error(FSComp.SR.buildErrorOpeningBinaryFile(filename, e.Message),m))
                 else 
                     tcImports.PrepareToImportReferencedILAssembly (ctok, m, filename, dllinfo)
-            dllinfo,phase2
+            return dllinfo, phase2
+      }
 
-    member tcImports.RegisterAndImportReferencedAssemblies (ctok, nms:AssemblyResolution list) =
+    member tcImports.RegisterAndImportReferencedAssemblies (ctok, nms: AssemblyResolution list) =
+      async {
         CheckDisposed()
 
-        let dllinfos,phase2s = 
-           nms |> List.choose 
-                    (fun nm ->
+        let! r =
+           nms 
+           |> Async.List.choose 
+                (fun nm ->
+                    async {
                         try
-                            Some(tcImports.RegisterAndPrepareToImportReferencedDll (ctok, nm))
+                            let! r = tcImports.RegisterAndPrepareToImportReferencedDll (ctok, nm)
+                            return Some r
                         with e ->
                             errorR(Error(FSComp.SR.buildProblemReadingAssembly(nm.resolvedPath, e.Message),nm.originalReference.Range))
-                            None)
-               |> List.unzip
+                            return None
+                    })
+        let  dllinfos, phase2s = List.unzip r
         let ccuinfos = (List.collect (fun phase2 -> phase2()) phase2s) 
-        dllinfos,ccuinfos
+        return dllinfos, ccuinfos
+      }
       
     member tcImports.DoRegisterAndImportReferencedAssemblies(ctok, nms) = 
```

Comment:
Done. However, this propagation stops now here https://github.com/Microsoft/visualfsharp/pull/2454/files#diff-7cfe8057cee7d7428672d69982354073R1616
I don't know what to do with it. Looks like we cannot push it forward without merging `eventually` with `async`.